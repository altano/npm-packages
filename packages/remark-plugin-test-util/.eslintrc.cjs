module.exports = {
  extends: ["altano"],
  overrides: [
    {
      files: ["src/**/*.ts", "src/**/*.tsx", "*.config.ts"],
      parserOptions: {
        project: ["./tsconfig.json"],
        tsconfigRootDir: __dirname,
      },
    },
  ],
};
