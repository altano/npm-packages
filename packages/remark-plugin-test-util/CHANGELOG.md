# @altano/remark-plugin-test-util

## 0.0.2

### Patch Changes

- 5fc1078: update dependencies
