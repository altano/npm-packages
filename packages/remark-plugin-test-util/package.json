{
  "name": "@altano/remark-plugin-test-util",
  "version": "0.0.1",
  "private": true,
  "description": "Testing utilities for remark plugins",
  "type": "module",
  "sideEffects": false,
  "main": "./index.ts",
  "scripts": {
    "lint": "eslint src/**/*.ts*",
    "lint:timing": "TIMING=1 eslint src/**/*.ts*",
    "lint:fix": "TIMING=1 eslint src/**/*.ts* --fix",
    "format": "prettier --write src",
    "clean": "rm -rf .turbo && rm -rf .tsbuildinfo && rm -rf node_modules && rm -rf dist"
  },
  "author": {
    "name": "Alan Norbauer",
    "email": "altano@gmail.com",
    "url": "https://alan.norbauer.com"
  },
  "license": "ISC",
  "repository": {
    "type": "git",
    "url": "https://github.com/altano/npm-packages",
    "directory": "packages/remark-plugin-test-util"
  },
  "homepage": "https://github.com/altano/npm-packages/tree/main/packages/remark-plugin-test-util",
  "devDependencies": {
    "@altano/tsconfig": "workspace:*",
    "vfile": "^5.3.7",
    "@mdx-js/mdx": "^2.3.0",
    "@types/node": "^18.15.13",
    "@types/prettier": "^2.7.2",
    "@vitest/ui": "^0.30.1",
    "eslint": "^8.39.0",
    "eslint-config-altano": "workspace:*",
    "mdast-util-mdxjs-esm": "^1.3.1",
    "prettier": "^2.8.7",
    "tsup": "^7.2.0",
    "typescript": "^5.1.6",
    "unified": "^10.1.2",
    "vitest": "^0.30.1"
  },
  "publishConfig": {
    "access": "restricted"
  }
}
