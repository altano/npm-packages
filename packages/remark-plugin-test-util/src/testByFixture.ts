import { compile } from "@mdx-js/mdx";
import { promises as fs } from "node:fs";
import { join } from "node:path";
import { expect, test } from "vitest";
import { VFile } from "vfile";

import type { Plugin } from "unified";

function isTodo(fixture: string): boolean {
  return fixture.startsWith("todo-");
}

function isError(fixture: string): boolean {
  return fixture.startsWith("error-");
}

export async function testByFixtures(plugin: Plugin): Promise<void> {
  const fixturesDir = join("tests", "unit", "__fixtures__");
  const fixtures = await fs.readdir(fixturesDir);
  fixtures.filter(isTodo).forEach((f) => test.todo(f));
  const rest = fixtures.filter((f) => !isTodo(f));
  test.each(rest)("[test #%#] %s", async (fixture) => {
    const compileWithPlugin = await getFixtureCompiler(plugin, fixture);
    if (isError(fixture)) {
      await expect(compileWithPlugin).rejects.toThrowErrorMatchingSnapshot();
    } else {
      const vfile = await compileWithPlugin();
      expect(vfile).toMatchSnapshot();
    }
  });
}

export async function getFixtureCompiler(
  plugin: Plugin,
  fixture: string,
  optionOverrides: object = {},
): Promise<() => Promise<VFile>> {
  const fixturesDir = join("tests", "unit", "__fixtures__");
  const fixtureDir = join(fixturesDir, fixture);
  const inputMdx = await fs.readFile(join(fixtureDir, "input.mdx"));
  const inputMdxPath = join(fixtureDir, `input.mdx`);
  const fixtureOptions = (await import(join(fixtureDir, "options.js"))).default;
  const options = {
    ...fixtureOptions,
    ...optionOverrides,
  };
  const compileWithPlugin = async (): Promise<VFile> => {
    const result = await compile(
      new VFile({ path: inputMdxPath, value: inputMdx }),
      {
        format: "mdx",
        remarkPlugins: [[plugin, options]],
        jsx: true,
      },
    );
    // The mdx compiler doesn't change the vfile path extension so let's fix
    // that. See https://github.com/mdx-js/mdx/issues/2462
    result.path = result.path.replace(/.mdx$/, ".js");
    return result;
  };

  return compileWithPlugin;
}
