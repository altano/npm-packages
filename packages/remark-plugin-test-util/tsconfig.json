{
  "extends": "@altano/tsconfig/node.json",
  "include": ["src"]
}
