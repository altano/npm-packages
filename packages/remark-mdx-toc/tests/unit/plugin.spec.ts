import plugin from "../../src/index.js";
import { describe } from "vitest";
import { testByFixtures } from "@altano/remark-plugin-test-util/testByFixture";

describe("remark-mdx-toc", async () => {
  await testByFixtures(plugin);
});
