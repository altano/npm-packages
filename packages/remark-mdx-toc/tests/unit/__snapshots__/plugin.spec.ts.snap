// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`remark-mdx-toc > [test #0] basic 1`] = `
FormattedVFile {
  "cwd": "<cwd>",
  "data": {},
  "history": [
    "tests/unit/__fixtures__/basic/input.mdx",
    "tests/unit/__fixtures__/basic/input.js",
  ],
  "map": undefined,
  "messages": [],
  "value": "/*@jsxRuntime automatic*/
/*@jsxImportSource react*/
export const toc = [
  {
    depth: 1,
    value: "Hello, world",
    attributes: {},
    children: [
      {
        depth: 2,
        value: "Title 1",
        attributes: {},
        children: [
          {
            depth: 3,
            value: "Subtitle 1",
            attributes: {},
            children: [],
          },
        ],
      },
      {
        depth: 2,
        value: "Title 2",
        attributes: {},
        children: [],
      },
    ],
  },
  {
    depth: 1,
    value: "Heading with html",
    attributes: {
      id: "h1",
    },
    children: [],
  },
];
function _createMdxContent(props) {
  const _components = {
    h1: "h1",
    h2: "h2",
    h3: "h3",
    p: "p",
    ...props.components,
  };
  return (
    <>
      <_components.h1>{"Hello, world"}</_components.h1>
      {"\\n"}
      <_components.h2>{"Title 1"}</_components.h2>
      {"\\n"}
      <_components.p>{"Content 1"}</_components.p>
      {"\\n"}
      <_components.h3>{"Subtitle 1"}</_components.h3>
      {"\\n"}
      <_components.p>{"Sub Content 1"}</_components.p>
      {"\\n"}
      <_components.h2>{"Title 2"}</_components.h2>
      {"\\n"}
      <_components.p>{"Content 2"}</_components.p>
      {"\\n"}
      <h1 id="h1">{"Heading with html"}</h1>
    </>
  );
}
export default function MDXContent(props = {}) {
  const { wrapper: MDXLayout } = props.components || {};
  return MDXLayout ? (
    <MDXLayout {...props}>
      <_createMdxContent {...props} />
    </MDXLayout>
  ) : (
    _createMdxContent(props)
  );
}",
}
`;

exports[`remark-mdx-toc > [test #1] custom-component 1`] = `
FormattedVFile {
  "cwd": "<cwd>",
  "data": {},
  "history": [
    "tests/unit/__fixtures__/custom-component/input.mdx",
    "tests/unit/__fixtures__/custom-component/input.js",
  ],
  "map": undefined,
  "messages": [],
  "value": "/*@jsxRuntime automatic*/
/*@jsxImportSource react*/
export const toc = [
  {
    depth: 1,
    value: "Heading with html",
    attributes: {
      id: "h1",
    },
    children: [
      {
        depth: 2,
        value: "Parented to HTML Heading Correctly",
        attributes: {},
        children: [],
      },
    ],
  },
  {
    depth: 1,
    value: "Heading with custom component",
    attributes: {},
    children: [
      {
        depth: 2,
        value: "Parented to Custom Component Heading Correctly",
        attributes: {},
        children: [],
      },
      {
        depth: 2,
        value: "Custom Sub-Heading",
        attributes: {},
        children: [],
      },
    ],
  },
];
function _createMdxContent(props) {
  const _components = {
      h2: "h2",
      p: "p",
      ...props.components,
    },
    { H1, H2 } = _components;
  if (!H1) _missingMdxReference("H1", true);
  if (!H2) _missingMdxReference("H2", true);
  return (
    <>
      <_components.p>
        {"const H2 = ("}
        {children}
        {") => "}
        <h2>{children}</h2>
      </_components.p>
      {"\\n"}
      <h1 id="h1">{"Heading with html"}</h1>
      {"\\n"}
      <_components.h2>{"Parented to HTML Heading Correctly"}</_components.h2>
      {"\\n"}
      <H1>{"Heading with custom component"}</H1>
      {"\\n"}
      <_components.h2>
        {"Parented to Custom Component Heading Correctly"}
      </_components.h2>
      {"\\n"}
      <H2>{"Custom Sub-Heading"}</H2>
    </>
  );
}
export default function MDXContent(props = {}) {
  const { wrapper: MDXLayout } = props.components || {};
  return MDXLayout ? (
    <MDXLayout {...props}>
      <_createMdxContent {...props} />
    </MDXLayout>
  ) : (
    _createMdxContent(props)
  );
}
function _missingMdxReference(id, component) {
  throw new Error(
    "Expected " +
      (component ? "component" : "object") +
      " \`" +
      id +
      "\` to be defined: you likely forgot to import, pass, or provide it.",
  );
}",
}
`;

exports[`remark-mdx-toc > [test #2] error-invalid-identifier 1`] = `[Error: Invalid name for an identifier: 1-invalid-identifier-starting-with-number]`;

exports[`remark-mdx-toc > [test #3] fragment 1`] = `
FormattedVFile {
  "cwd": "<cwd>",
  "data": {},
  "history": [
    "tests/unit/__fixtures__/fragment/input.mdx",
    "tests/unit/__fixtures__/fragment/input.js",
  ],
  "map": undefined,
  "messages": [],
  "value": "/*@jsxRuntime automatic*/
/*@jsxImportSource react*/
export const toc = [
  {
    depth: 1,
    value: "Hello, world",
    attributes: {},
    children: [],
  },
  {
    depth: 1,
    value: "Goodbye, world",
    attributes: {},
    children: [],
  },
];
function _createMdxContent(props) {
  const _components = {
    h1: "h1",
    ...props.components,
  };
  return (
    <>
      <_components.h1>{"Hello, world"}</_components.h1>
      {"\\n"}
      <>{"Some fragment"}</>
      {"\\n"}
      <_components.h1>{"Goodbye, world"}</_components.h1>
    </>
  );
}
export default function MDXContent(props = {}) {
  const { wrapper: MDXLayout } = props.components || {};
  return MDXLayout ? (
    <MDXLayout {...props}>
      <_createMdxContent {...props} />
    </MDXLayout>
  ) : (
    _createMdxContent(props)
  );
}",
}
`;

exports[`remark-mdx-toc > [test #4] non-heading-flow-element 1`] = `
FormattedVFile {
  "cwd": "<cwd>",
  "data": {},
  "history": [
    "tests/unit/__fixtures__/non-heading-flow-element/input.mdx",
    "tests/unit/__fixtures__/non-heading-flow-element/input.js",
  ],
  "map": undefined,
  "messages": [],
  "value": "/*@jsxRuntime automatic*/
/*@jsxImportSource react*/
export const toc = [];
function _createMdxContent(props) {
  const _components = {
    em: "em",
    p: "p",
    ...props.components,
  };
  return (
    <>
      <em>{"I'm in a non-heading, mdx flow element"}</em>
      {"\\n"}
      <_components.p>
        <_components.em>
          {"I'm in a non-heading, markdown flow element"}
        </_components.em>
      </_components.p>
    </>
  );
}
export default function MDXContent(props = {}) {
  const { wrapper: MDXLayout } = props.components || {};
  return MDXLayout ? (
    <MDXLayout {...props}>
      <_createMdxContent {...props} />
    </MDXLayout>
  ) : (
    _createMdxContent(props)
  );
}",
}
`;

exports[`remark-mdx-toc > [test #5] use-default-name 1`] = `
FormattedVFile {
  "cwd": "<cwd>",
  "data": {},
  "history": [
    "tests/unit/__fixtures__/use-default-name/input.mdx",
    "tests/unit/__fixtures__/use-default-name/input.js",
  ],
  "map": undefined,
  "messages": [],
  "value": "/*@jsxRuntime automatic*/
/*@jsxImportSource react*/
export const toc = [
  {
    depth: 2,
    value: "Resource Acquisition Is Initialization (RAII)",
    attributes: {},
    children: [
      {
        depth: 3,
        value: "Is There a Problem?",
        attributes: {},
        children: [],
      },
      {
        depth: 3,
        value: "C++ Classes",
        attributes: {},
        children: [],
      },
      {
        depth: 3,
        value: "The Solution",
        attributes: {},
        children: [],
      },
    ],
  },
];
function _createMdxContent(props) {
  const _components = {
    h2: "h2",
    h3: "h3",
    ...props.components,
  };
  return (
    <>
      <_components.h2>
        {"Resource Acquisition Is Initialization (RAII)"}
      </_components.h2>
      {"\\n"}
      <_components.h3>{"Is There a Problem?"}</_components.h3>
      {"\\n"}
      <_components.h3>{"C++ Classes"}</_components.h3>
      {"\\n"}
      <_components.h3>{"The Solution"}</_components.h3>
    </>
  );
}
export default function MDXContent(props = {}) {
  const { wrapper: MDXLayout } = props.components || {};
  return MDXLayout ? (
    <MDXLayout {...props}>
      <_createMdxContent {...props} />
    </MDXLayout>
  ) : (
    _createMdxContent(props)
  );
}",
}
`;
